docs/async.md

```diff
@@ -0,0 +1,73 @@
+# ASYNC
+
+- [Description](#description)
+- [Goals](#goals)
+- [Code](#code)
+- [Limitations](#limitations)
+
+
+## Description
+
+Apisonator can be configured to use a non-blocking redis client. It's an opt-in
+feature that can be enabled setting the `CONFIG_REDIS_ASYNC` env to `true`. When
+the feature is enabled, Apisonator uses the
+[async-redis](https://github.com/socketry/async-redis) client instead of
+[redis-rb](https://github.com/redis/redis-rb).
+
+
+## Goals
+
+Using a non-blocking client should allow us to save resources. Also it makes it
+easier to decide how to scale. With a blocking client it's difficult to decide
+how many listeners/workers should be deployed per machine in order to use all
+the available CPU. With a non-blocking client, in theory, a single process
+should use a whole CPU when properly tuned.
+
+
+## Code
+
+We made a few changes in the codebase to support this feature. Here are the most
+important ones:
+
+- There are two worker classes. `WorkerAsync` and `WorkerSync`. `WorkerAsync`
+needs to run a separate thread to fetch jobs from the queues. Apart from that,
+they are pretty much the same.
+
+- There's `Backend::StorageAsync::Client` and `Backend::StorageSync`.
+`Backend::Storage` is now responsible for instantiating the correct type of
+client based on the configuration.
+
+- The majority of the code required to make async work is under
+`lib/3scale/backend/storage_async`. That includes a wrapper for the
+`async-redis` client, and some monkey-patching to optimize a few things. Some
+tests helpers also needed to be adapted to choose between the two types of
+clients.
+
+
+## Limitations
```

Comment:
There is no mention of what the async gem does, and it is a potential source of problems. In particular, it monkey-patches core IO modules to use a reactor, which means that interaction with any IO code is to be suspected in case of issues, and C extensions (usually) won't use the reactor for their IO (so we need to watch out for blocking there).

Reply:
It doesn’t do that (monkey patching).

Reply:
that was my understanding from when the first PR was introduced - are you not modifying the behaviour of `IO` methods?

I'd then assume async-redis is using async{,-io} on its own and then nothing else would be touching the reactor, right? That is preferable for us as it lowers the exposure to issues in unrelated code.

Reply:
> that was my understanding from when the first PR was introduced - are you not modifying the behaviour of IO methods?

@unleashed we are not modifying IO. We add `Async::IO` which provides (almost) identical wrappers which you can inject into other code if needed.

> I'd then assume async-redis is using async{,-io} on its own and then nothing else would be touching the reactor, right? That is preferable for us as it lowers the exposure to issues in unrelated code.

At the moment, that's correct. However in the future, we may make it possible for the user to allow other forms of IO to use the reactor: https://github.com/ruby/ruby/pull/1870 - just FYI.